Bugfix/bb 286 fix kafka log consumer not letting log reader push to topics

welansari opened the pull request:
Issue: [BB-286](https://scality.atlassian.net/browse/BB-286)

**Original Issue:**
Previous implementation of async in the logReader produces flakiness where some oplog entries can get processed by the filter function but not pushed into their respective topic.
The issue is only present for backbeat extensions that use an async filter function (only Bucket Notification for now).

**Changes:**
- Implement proper support for async `filter` functions (same as in 7.x branch)
- Stringified the oplog event value (bug)
- Implement filterAsync in the NotificationQueuePopulator
- Avoid crashing when having an issue consuming messages from kafka. This was done to avoid getting crashes at startup before the oplog topic get created.

Zenko Build: https://github.com/scality/Zenko/actions/runs/3376337223

Commit:
bugfix: BB-286 stringify oplog entry value

Backbeat extensions expect the entry.value field to be
a stringified JSON containing the metadata of the object

Commit:
bugfix: BB-286 make logReader support async filter functions

Some extensions like Bucket Notification have to use asynchronous filter
functions. This commit adds a proper support for that in the logReader
while still maintaining the callback style functions that are used everywhere
in the file.

Commit:
bugfix: BB-286 implement filterAsync in the notificationQueuePopulator

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-286 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.21`


Please check the `Fix Version/s` of BB-286, or the target
branch of this pull request.

Commit:
bugfix: BB-286 avoid crashing when getting error consuming kafka messages

Kafka topics are not created at startup, but only once the first message
is pushed into the topics.

We avoid crashing backbeat when the kafka logConsumer is  having an issue consuming
a message from a specific topic, the operation will be retried anyway in the next
batch read opetation.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-286 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.22`


Please check the `Fix Version/s` of BB-286, or the target
branch of this pull request.

francoisferrand commented on lib/queuePopulator/LogReader.js:
```diff
@@ -522,6 +484,94 @@ class LogReader {
+            if (typeof ext.filterAsync === 'function') {
+                ext.filterAsync(entry, cb);
+            } else {
+                ext.filter(entry);
+                process.nextTick(cb);
```
this should be just `callbackify(ext.filter)`

francoisferrand replied:
also, wondering if this should not be done just once, e.g. on startup process the list of extenions to build the list of callbacks : then just call them for processing

welansari replied:
IMO i think that adds unnecessary complexity to the code especially since we are not waiting on any return value

francoisferrand commented on lib/queuePopulator/LogReader.js:
```diff
@@ -522,6 +484,94 @@ class LogReader {
+            if (typeof ext.filterAsync === 'function') {
+                ext.filterAsync(entry, cb);
+            } else {
+                ext.filter(entry);
+                process.nextTick(cb);
```
can we make this more seamless and support async methods, by just checking the return?

e.g. maybe something like this:
```
res = ext.filter(entry)
if (res instanceof Promise) {
   res.then(cb)
} else {
   process.nextTick(cb);
}
```

And then maybe we do not even need to switch to callbacks in the extensions (or support filterAsync)....

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-286 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.24`

* `8.5.0`


Please check the `Fix Version/s` of BB-286, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-286 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.24`

* `8.5.1`


Please check the `Fix Version/s` of BB-286, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-286 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.25`

* `8.5.2`


Please check the `Fix Version/s` of BB-286, or the target
branch of this pull request.

williamlardier commented on lib/queuePopulator/LogReader.js:
```diff
@@ -522,6 +484,94 @@ class LogReader {
+                entry.type === 'delete' &&
+                ext.constructor.name === 'NotificationQueuePopulator'
+            ) {
+                entry.value = JSON.stringify({
```
Same comment here for the JSON.stringify

welansari replied:
No need i think inside are values we set

https://github.com/scality/Arsenal/blob/c343820cae2c7a533fb73046b2ad4c13cb77fc8b/lib/storage/metadata/mongoclient/ListRecordStream.js#L133-L137

williamlardier commented on lib/queuePopulator/LogReader.js:
```diff
@@ -522,6 +484,94 @@ class LogReader {
+            this._unsetEntryBatch();
+            if (err) {
+                this.log.error('error filtering entries from log', {
+                    method: 'LogReader._processFilterEntries',
+                    error: err,
```
Can it be better to include `batchState` in the error log to ease the debugging? (same for other logs)

welansari replied:
A batch can contain up to 10 000 events, i'm afraid logging the `batchState` would just give too much information that might not even help when debugging.

One thing to note here also is that at the end of each batch we log the `batchState.logStat` which contains information about the events we read
https://github.com/scality/backbeat/blob/db9e3815983b046b77bc41847dbf201006b20d0a/lib/queuePopulator/LogReader.js#L311-L317

williamlardier commented on lib/queuePopulator/LogReader.js:
```diff
@@ -522,6 +484,94 @@ class LogReader {
+
+    _processFilterEntries(batchState, done) {
+        const { currentRecords } = batchState;
+        if (!currentRecords || currentRecords.length === 0) {
```
Maybe safer like that, in case length is not defined?

```suggestion
        if (!currentRecords || !currentRecords.length) {
```

williamlardier commented on lib/queuePopulator/LogReader.js:
```diff
@@ -522,6 +484,94 @@ class LogReader {
+
+    _processFilterEntry(batchState, record, done) {
+        const { entriesToPublish } = batchState;
+        if (!record || !record.entries) {
+            return done();
+        }
```
Can be done before line 537

williamlardier reviewed (commented):
Some code-related comments.

The logic LGTM as well, I did not see anything wrong according to your ticket objectives.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-286-fix-kafka-logConsumer-not-letting-logReader-push-to-topics` into
`development/8.4`
* [w/8.5/bugfix/BB-286-fix-kafka-logConsumer-not-letting-logReader-push-to-topics](https://github.com/scality/backbeat/compare/development/8.5...w/8.5/bugfix/BB-286-fix-kafka-logConsumer-not-letting-logReader-push-to-topics) will be merged into `development/8.5` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
skip event in kafka logConsumer when event object is not valid

bert-e commented:
# History mismatch



Merge commit #ef12049ba646a5d5d4a576722345e37f6cd8e686 on the integration branch
`w/8.5/bugfix/BB-286-fix-kafka-logConsumer-not-letting-logReader-push-to-topics` is merging a branch which is neither the current
branch `bugfix/BB-286-fix-kafka-logConsumer-not-letting-logReader-push-to-topics` nor the development branch
`development/8.5`.

It is likely due to a rebase of the branch `bugfix/BB-286-fix-kafka-logConsumer-not-letting-logReader-push-to-topics` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

welansari commented on lib/queuePopulator/KafkaLogConsumer/ListRecordStream.js:
```diff
@@ -65,7 +65,7 @@ class ListRecordStream extends stream.Transform {
                 key: changeStreamDocument.documentKey && changeStreamDocument.documentKey._id,
                 type: this._getType(changeStreamDocument.operationType),
-                value: changeStreamDocument.fullDocument && changeStreamDocument.fullDocument.value,
+                value: changeStreamDocument.fullDocument && JSON.stringify(changeStreamDocument.fullDocument.value),
```
We already parsed the string before, so i don't see a reason why it can't be re-parsed again. If the object happens to be circular the first try/catch should handle the error

The thing to change here i think is the way we handle those errors, i think returning an empty event is irrelevant as it will be ignored anyway, i changed it to skip the event completely

williamlardier replied:
I see. Now my question is: do we really need to parse the same string several time?

welansari replied:
The thing here is that we can not parse a part of the string. we only need the fullDocument field as a string (this is where the object metadata lies)

welansari replied:
Oh i just noticed, i said re-parsed when i should have said stringified 😅.
The idea is if it got parsed with no issue then it should stringify with no errors

williamlardier reviewed (approved):
LGTM

welansari commented:
/approve

bert-e commented:
# History mismatch



Merge commit #ef12049ba646a5d5d4a576722345e37f6cd8e686 on the integration branch
`w/8.5/bugfix/BB-286-fix-kafka-logConsumer-not-letting-logReader-push-to-topics` is merging a branch which is neither the current
branch `bugfix/BB-286-fix-kafka-logConsumer-not-letting-logReader-push-to-topics` nor the development branch
`development/8.5`.

It is likely due to a rebase of the branch `bugfix/BB-286-fix-kafka-logConsumer-not-letting-logReader-push-to-topics` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

welansari commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-286-fix-kafka-logConsumer-not-letting-logReader-push-to-topics` into
`development/8.4`
* [w/8.5/bugfix/BB-286-fix-kafka-logConsumer-not-letting-logReader-push-to-topics](https://github.com/scality/backbeat/compare/development/8.5...w/8.5/bugfix/BB-286-fix-kafka-logConsumer-not-letting-logReader-push-to-topics) will be merged into `development/8.5` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/8.3`



Please check the status of the associated issue BB-286.

Goodbye kerkesni.
